csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/tests/Tests.kt

```diff
@@ -0,0 +1,289 @@
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are duplicate: ${duplicates.keys.joinToString(",")}")
+            )
+        }
+    }
+}
+
+private fun <T> List<T>.duplicates(): Map<T, Int> {
+    return groupingBy { it }.eachCount().filter { it.value > 1 }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.3](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#613-circular-definition-of-product-id).
+ */
+object Test613CircularDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val circles = mutableSetOf<String>()
+
+        for (relationship in doc.product_tree?.relationships ?: listOf()) {
+            val definedId = relationship.full_product_name.product_id
+            val notAllowed =
+                listOf(relationship.product_reference, relationship.relates_to_product_reference)
+            if (definedId in notAllowed) {
+                circles += definedId
+            }
+        }
+
+        return if (circles.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are defined in circles: ${circles.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.4](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#614-missing-definition-of-product-group-id).
+ */
+object Test614MissingDefinitionOfProductGroupID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProductGroups().map { it.group_id }
+        val references = doc.gatherProductGroupReferences()
+
+        val notDefined = references.subtract(definitions.toSet())
+
+        return if (notDefined.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are not defined: ${notDefined.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.5](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#615-multiple-definition-of-product-group-id)
+ */
+object Test615MultipleDefinitionOfProductGroupID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProductGroups().map { it.group_id }
+
+        val duplicates = definitions.duplicates()
+
+        return if (duplicates.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are duplicate: ${duplicates.keys.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.6](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#616-contradicting-product-status).
+ */
+object Test616ContradictingProductStatus : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val affected = mutableSetOf<String>()
```

Comment:
I'd move `affected`, `notAffected`, `fixed`, `underInvestigation` inside the loop instead of having them here, calling clear all the time and so on. There's no need to keep the information outside the loop